haddock-api/src/Haddock/GhcUtils.hs

```diff
@@ -226,6 +228,95 @@ getGADTConTypeG (ConDeclH98 {}) = panic "getGADTConTypeG"
   -- Should only be called on ConDeclGADT
 getGADTConTypeG (XConDecl {}) = panic "getGADTConTypeG"
 
+
+-------------------------------------------------------------------------------
+-- * Parenthesization 
+-------------------------------------------------------------------------------
+
+-- | Precedence level (inside the 'HsType' AST).
+data Precedence
```

Comment:
Perhaps we could reuse the [`PprPrec` data type](http://git.haskell.org/ghc.git/blob/45f0026818402aa08398131507bc587ea4a2b387:/compiler/basicTypes/BasicTypes.hs#l674) from GHC itself?

Reply:
GHC has `funPrec == opPrec`, while Haddock has `funPrec < opPrec`. The `Note [Type operator precedence]` summarizes this difference quite well: GHC's choice of precedence is not incorrect - it will just lead to a extra parens.

In summary I'd rather not, but I'm willing to be told otherwise.

Reply:
Ah, I missed that detail. In that case, what you've used here is fine.